### DISCUSSION TITLE
[Question] Is there a reason why Job.id is marked as optional in type definitions?

### DISCUSSION THREAD

**Discussion by BOTKooper** [taskforcesh/bullmq#2547] [Q&A] [answered]
I need to sync my db with bull a bit to track jobs, but while doing that i noticed that `Job.id` is marked as optional in both typescript definitions as well as at [official docs](https://api.docs.bullmq.io/classes/v5.Job.html#id).

Is that a bug in documentation, or there is a edge-case when worker can actually get a job without one?

I can see that it actually doesn't have one before scheduler creates it, but that coming into _Worker_ type-def sounds a bit silly

P.S. not a dealbraker or anything - I still can pass it as part of the `Job.data` in my case, just a bit unexpected

**Comment by manast** [comment]
The reason is that a job does not get an id until it is stored in Redis... maybe there were more elegant ways to design this API, but I think it is difficult to change it now without a breaking change.

**Reply by BOTKooper** [reply]
That's reasonable, though I do not know if bullmq can skip redis when giving job to worker (which would be _surprising_ to say the least).

If that's not the case - maybe we can make it smth like `JobWithId` inside Processor params? Shouldn't be a breaking change, but rather a clarification

**Reply by manast** [reply, marked as answer]
I think in that case what we can do is add a type guard to the process function where we know the id is present:
```Job & { id: string }```
